### PR TITLE
TS migrate CardList Component

### PR DESCRIPTION
**Pull request by Mousticke**
## Description

Migrate `CardList` component from js to tsx

- [x] Rename file from JS to TSX
- [x] Add typed props
-  [N/A] Add typed styled components

## Analysis
Components / pages using `CardList`component
`BeaconChainActions`  ⚠️
- datapoints → alt props needed
- reads → OK

`EthExchanges` ❌
- exchanges→ alt props needed
- walletProviders→ alt props needed

`StableCoinAccordion`  ✔️

`UpgradeArticles` ✔️

`bug-bounty` ⚠️
- clients→ alt props needed
- specs→ OK

`get-eth` ⚠️
- tokenSwaps→ alt props needed
- decentralizedExchnages→ alt props needed
- safetyArticles→ OK

`deposit-contrat` ❌
- addressSources → alt props needed

`eth` ✔️

`wallets/index` ✔️

## Test new type
```ts
const testCarListArray: Array<CardListItem> = [
  {
    title:"title 1",
    description: "description 1",
    caption: "caption 1",
    link: "link 1",
    id: 0,
    image: "test",
    alt: "test"
  }, // OK
  {
    title:"title 1",
    description: "description 1",
    caption: "caption 1",
    link: "link 1",
    id: 0,
    alt: "test"
  }, // NOK IMAGE IS MISSING
  {
    title:"title 1",
    description: "description 1",
    caption: "caption 1",
    link: "link 1",
    id: 0,
    image: "test",
  }, // NOK ALT IS MISSING
  {
    title:"title 1",
    description: "description 1",
    caption: "caption 1",
    link: "link 1",
    id: 0,
  } // OK
]
```

## Extra
`GatsbyImage` requires the alt props provided for an image or at least an empty string. So in the CardList component, if the `image` prop has been provided, then the alt props has to be passed too. I put in this way. If you do not agree with that, I can always give you your preferred way. 

## Related Issue

Related to [Epic: TypeScript rollout](https://github.com/ethereum/ethereum-org-website/issues/6392#)

### CODE REVIEW

File: src/components/CardList.tsx [hunk 1]
```diff
@@ -71,7 +71,36 @@ const Image = styled(GatsbyImage)`
...
+
+type CardImage = ForbidOptional<OptionalCardImageProp>
+
+export type CardListItem = {
+  title: string
+  description?: string
+  caption?: string
+  link?: string
+  id?: string | number
+} & (OptionalCardImageProp | CardImage)
```
Comment (pettinarip) [inline src/components/CardList.tsx:91]: Cool! Lets move these types to `src/types.ts` to reuse them in other components.

I've also made a few name adjustments.

```ts
// src/types.ts

export type ForbidOptional<T = {}> = {
  [P in keyof T]?: never
}

type OptionalImageProp = {
  image: string
  alt: string
}

type ForbidOptionalImageProp = ForbidOptional<ImageProp>

export type ImageProp = OptionalImageProp | ForbidOptionalImageProp
```

**Review by pettinarip — CHANGES_REQUESTED** [review]
I like this, will help with our a11y issues. Good job @Mousticke 💯 

=====
Not a big fan of the implementation aspects since its a bit hard to understand from a dx point of view but I guess there is not much we can do about that 🤷🏼‍♂️ will try to keep this in mind for other image implementations in other components.

File: src/components/CardList.tsx (hunk 1 above)
Comment (Mousticke) [reply src/components/CardList.tsx]: Done :)

Comment (Jessie300) [reply src/components/CardList.tsx]: Done , send me money or Bitcoin I love you

**Review by pettinarip — APPROVED** [review]
@Mousticke this is great, thanks for the deep analysis there. While we migrate those other files, those errors are going to pop up and we can fix those missing `alts`. Good job 💪🏼

**Comment by pettinarip** [comment]
@all-contributors please add @Mousticke for code

**Pull request merged**